### PR TITLE
[Fixes 1428] Setting ContentEncoding header in ContentResult

### PR DESCRIPTION
**Pull request by ajaybhargavb**
@kichalla

### CODE REVIEW

**Comment by yishaigalatzer** [comment]
@Tratcher are strongly typed headers (and particularly support for content type/char set) is going to be available early this week?

**Comment by Tratcher** [comment]
That's the goal. I'll send the PR tomorrow.

Sent from my Windows Phone
________________________________

**Comment by ajaybhargavb** [comment]
Updated. Changes based on https://github.com/aspnet/HttpAbstractions/pull/164.

File: src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs [hunk 1]
```diff
@@ -21,7 +22,14 @@ public class ContentResult : ActionResult
 
             if (!string.IsNullOrEmpty(ContentType))
             {
-                response.ContentType = ContentType;
+                var contentTypeHeader = new MediaTypeHeaderValue(ContentType);
+                contentTypeHeader.Encoding = ContentEncoding;
+                response.ContentType = contentTypeHeader.ToString();
+            }
+
+            if (ContentEncoding != null)
```
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs]: you can remove this now as you are already setting this above

File: test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/ContentResultTest.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+        private static HttpContext GetHttpContext()
+        {
+            var httpContext = new Mock<HttpContext>();
+            var realContext = new DefaultHttpContext();
+            var response = realContext.Response;
+            httpContext.Setup(o => o.Response)
+                       .Returns(response);
+
+            return httpContext.Object;
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/ContentResultTest.cs]: You can just use `return new DefaultHttpContext()` here. It's supposed to be usable in tests.

**Comment by rynowak** [comment]
:watch: for a functional test. 

/cc @Tratcher

File: src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs [hunk 3]
```diff
@@ -21,7 +22,9 @@ public class ContentResult : ActionResult
 
             if (!string.IsNullOrEmpty(ContentType))
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs]: This should probably be MediaType, ContentType is the whole header.

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs]: #BackCompat

**Comment by rynowak** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs [hunk 4]
```diff
@@ -21,7 +22,9 @@ public class ContentResult : ActionResult
...
             {
-                response.ContentType = ContentType;
+                var contentTypeHeader = new MediaTypeHeaderValue(ContentType);
+                contentTypeHeader.Encoding = ContentEncoding;
```
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs]: ContentEncoding can be null if the user does not provide it...in this case should we default to UTF8 as that is what the HttpResponse object seems to be doing in the end? @Tratcher ...fyi:
https://github.com/aspnet/HttpAbstractions/blob/7be9cefc4c7671caba7d8fe538ba6a0ad308fdad/src/Microsoft.AspNet.Http/HttpResponseWritingExtensions.cs#L25

File: test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/ContentResultTest.cs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+using Microsoft.AspNet.Routing;
+using Moq;
+using Xunit;
+
+namespace Microsoft.AspNet.Mvc
+{
+    public class ContentResultTest
+    {
+        [Fact]
+        public async Task ContentResult_Response_SetsContentType()
```
Comment (kichalla) [inline test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/ContentResultTest.cs]: can you add a case where Content is null

File: src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs [hunk 6]
```diff
@@ -21,7 +22,9 @@ public class ContentResult : ActionResult
...
+                response.ContentType = contentTypeHeader.ToString();
             }
 
             if (Content != null)
```
Comment (kichalla) [inline src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs]: if Content is null we do not write the response, so should we return `Task.FromResult(true)` in this case... @Tratcher : will this be a non-complete task? and if yes, can it cause problems in hosts?...like my thought was that headers are sent the moment body is being written, but here we just set the headers and we do not write any content...so how would the host know to send the headers anyway...does it look for completed tasks?

File: src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs (hunk 4 above)
Comment (Tratcher) [inline src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs]: Yes, default to UTF-8. You should also be calling the overload of WriteAsync that takes an encoding or else UTF-8 will always be used.

File: src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs (hunk 6 above)
Comment (Tratcher) [inline src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs]: You're in an `async Task` method, the compiler will end the task implicitly at the end of the method. The server will send the headers at the end of the task chain if they haven't been sent yet.

**Comment by kichalla** [comment]
:shipit:

**Pull request merged**